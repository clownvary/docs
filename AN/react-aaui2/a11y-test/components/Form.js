import React from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'
import _ from 'lodash'

import {
  Form,
  NestedForm,
  TextInput,
  TextArea,
  UrlInput,
  EmailInput,
  Button,
  PhoneInput,
  NumericInput,
  Combobox,
  Radio,
  FormRadioGroup,
  AddressEditor,
  PostalCodeInput,
} from '../../src/index'
import FormTextArea from '../../src/form/TextArea'
import CountriesConfig from '../../src/form/config/countryConfig.json'
import L10nProvider from '../../src/shared/L10nProvider'
import { reduceInitFields } from '../../src/Form'

const INITIAL_FIELDS_DATA = {
  firstName: {
    value: '',
    parser: v => v,
    formatter: v => v,
    rules: 'required|min:10|max:10',
  },
  lastName: 'last name',
  sports: 'tennis',
  address: {
    fields: {
      country: 'ID',
      line1: 'line11111',
      line2: 'line22222',
      city: 'city11111\ncity2',
      stateProvince: 'AE',
      postalCode: '121212',
    },
  },
}
const localeStr = 'de_DE,de_AT,de_CH,en_AU,en_CA,en_GB,en_ZA,en_IE,en_NZ,en_US,es_ES,fr_CA,fr_FR,fr_CH,it_IT,it_CH,pt_BR,sv_SE,zh_CN'
const localeData = localeStr.split(',').map(l => ({
  text: l,
  value: l,
}))
const dropdownData = [{
  text: 'Tennis',
  value: 'tennis',
}, {
  text: 'Basketball',
  value: 'basketball',
}, {
  text: 'Swimming',
  value: 'swimming',
}]
const NestedComponent = ({ handleClearBtnClick }) => (
  <div>
    <hr />
    <Form.HField name='firstName' label='First Name' component={TextInput} required />
    <Form.HField name='new_sports' label='Sports' component={Combobox} data={dropdownData} required />
    <Button noSubmit onClick={handleClearBtnClick}>Empty Dynamic Fields</Button>
    <hr />
  </div>
)

NestedComponent.propTypes = {
  dynamicInsertFormFields: PropTypes.bool,
}

class App extends React.PureComponent {
  state = {
    fields: '',
    locale: 'en_US',
    country: 'can',
    richText: '',
    INITIAL_FIELDS: INITIAL_FIELDS_DATA,
  }

  handleCountryChange = ({ value }) => {
    this.setState({ country: value })
  }

  handleChange = fields => {
    this.setState({ fields: JSON.stringify(fields, null, 4) })
  }

  handleLocaleChange = ({ value }) => {
    this.setState({ locale: value })
  }

  handleEditorChange(e) {
    this.setState({ richText: e.target.getContent() })
  }

  handleUpdateFirstNameClick = () => {
    const { INITIAL_FIELDS } = this.state

    INITIAL_FIELDS.firstName = String((new Date()).valueOf())
    INITIAL_FIELDS.new_sports = 'basketball'

    this.setState({
      dynamicInsertFormFields: true,
      INITIAL_FIELDS: _.cloneDeep(INITIAL_FIELDS),
    })
  }

  handleClearBtnClick = e => {
    const { INITIAL_FIELDS } = this.state

    INITIAL_FIELDS.firstName = ''
    INITIAL_FIELDS.new_sports =  ''
    
    this.setState({
      INITIAL_FIELDS: _.cloneDeep(INITIAL_FIELDS),
    })
  }

  handleTextAreaChange = e => {
    this.setState({
      INITIAL_FIELDS: reduceInitFields(JSON.parse(e.target.value).values),
    })
  }

  render() {
    const { country, locale, INITIAL_FIELDS, dynamicInsertFormFields } = this.state

    return (
      <div style={{ fontFamily: 'ProximaNova, Arial, sans-serif' }}>
        <L10nProvider locale={locale}>
          <div className='grid'>
            <div className='grid-u-8-12'>
              <Form
                fields={INITIAL_FIELDS}
                onChange={this.handleChange}
                onSubmit={values => alert(JSON.stringify(values, null, 4))}
              >
                <Form.Error />
                <Form.HField name='locale' label='Locale' defaultValue={locale} component={Combobox} data={localeData} onChange={this.handleLocaleChange} />
                <hr />
                <Button noSubmit onClick={this.handleUpdateFirstNameClick}>Dynamic Insert Nested First Name and Sports Form Fields</Button>
                <hr />
                {dynamicInsertFormFields && <NestedComponent handleClearBtnClick={this.handleClearBtnClick} />}
                <Form.VField name='lastName' label='Last Name(VField)' rules='min:2|max:10' component={TextInput} required />
                <Form.HField name='email' label='Email' component={EmailInput} required />
                <Form.HField name='url' label='Url' component={UrlInput} />
                <Form.HField name='phone' label='Phone' component={PhoneInput} />
                <Form.HField name='number' label='Numeric' component={NumericInput} />
                <Form.HField name='regexp' label='RegExp (regexp:(^[a-z][0-9]$))' component={TextInput} rules='regexp:(^[a-z][0-9]$)' />
                <Form.HField name='postalCode' label='US Postal Code' component={PostalCodeInput} countryConfig={CountriesConfig.US} />
                <Form.HField name='combobox' label='combobox' component={Combobox} data={[{ text: 'TEST', value: 'test' }]} />
                <Form.HField name='description' label='Description' component={FormTextArea} rows={10} required />
                <Form.HField name='sports' label='Sports' component={FormRadioGroup}>
                  <Radio value='basketball'>Basketball</Radio>
                  <Radio value='tennis'>Tennis</Radio>
                  <Radio value='swimming'>Swimming</Radio>
                </Form.HField>
                <hr />
                <p>AddressEditor</p>
                <Form.Field name='address' component={AddressEditor} countriesConfig={CountriesConfig} />
                <hr />
                <Form.Field name='anotherAddress' component={NestedForm}>
                  <Form>
                    <p>Here is the nested form</p>
                    <Form.HField name='addressLine' label='addressForm.addressField.US.line1' required component={TextInput} />
                    <Form.HField
                      name='addressCountry'
                      onChange={this.handleCountryChange}
                      value={country}
                      label='addressForm.addressField.country'
                      component={Combobox}
                      data={[{ text: 'Canada', value: 'can' }, { text: 'China', value: 'chs' }, { text: 'Japan1', value: 'jap1' }]}
                      required
                    />
                  </Form>
                </Form.Field>
                <Button type='submit'>Submit</Button>
              </Form>
            </div>
            <div className='grid-u-4-12'>
              <p><strong>Displaying the values and errors here</strong></p>
              <TextArea rows='25' value={this.state.fields} onChange={this.handleTextAreaChange} />
              <Button noSubmit onClick={this.handleUpdateFirstNameClick}>
                Dynamic Insert Nested First Name and Sports Form Fields
              </Button>
            </div>
          </div>
        </L10nProvider>
      </div>
    )
  }
}


ReactDOM.render(<App />, document.getElementById('root'))
