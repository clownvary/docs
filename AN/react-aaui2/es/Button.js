import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import React from 'react';
import { bool, string, node } from 'prop-types';
import classNames from 'classnames';

var Button = function Button(_ref) {
  var _classNames;

  var noSubmit = _ref.noSubmit,
      loading = _ref.loading,
      type = _ref.type,
      size = _ref.size,
      children = _ref.children,
      className = _ref.className,
      rest = _objectWithoutProperties(_ref, ['noSubmit', 'loading', 'type', 'size', 'children', 'className']);

  var classes = classNames((_classNames = {
    btn: true
  }, _classNames['btn-' + type] = type, _classNames['btn--' + size] = size, _classNames['btn--loading'] = loading, _classNames), className);

  return React.createElement(
    'button',
    _extends({}, rest, { type: noSubmit ? 'button' : 'submit', className: classes }),
    loading && React.createElement('i', { className: 'icon-spinner icon--loading' }),
    React.createElement(
      'span',
      null,
      children
    )
  );
};

export default Button;

Button.displayName = 'AUIButton';
Button.propTypes = {
  noSubmit: bool,
  loading: bool,
  type: string,
  size: string,
  className: string,
  children: node
};

Button.defaultProps = {
  noSubmit: false,
  loading: false,
  type: 'secondary'
};