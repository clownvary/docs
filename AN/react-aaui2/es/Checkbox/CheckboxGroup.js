import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { PureComponent } from 'react';
import { bool, string, func, object, node, array, arrayOf, shape, any } from 'prop-types';
import classNames from 'classnames';

import Checkbox from './Checkbox';
import { identity } from '../shared/utils';

var CheckboxGroup = function (_PureComponent) {
  _inherits(CheckboxGroup, _PureComponent);

  function CheckboxGroup(props) {
    _classCallCheck(this, CheckboxGroup);

    var _this = _possibleConstructorReturn(this, _PureComponent.call(this, props));

    _initialiseProps.call(_this);

    var value = [];
    if ('value' in props) {
      value = props.value;
    } else if ('defaultValue' in props) {
      value = props.defaultValue;
    }

    _this.state = {
      value: value
    };
    return _this;
  }

  CheckboxGroup.prototype.getChildContext = function getChildContext() {
    return {
      auiCheckboxGroup: {
        onChange: this.onChange,
        value: this.state.value,
        disabled: this.props.disabled,
        size: this.props.size,
        name: this.props.name
      }
    };
  };

  CheckboxGroup.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    if ('value' in nextProps) {
      this.setState({
        value: nextProps.value
      });
    }
  };

  CheckboxGroup.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        data = _props.data;

    var classes = classNames('checkbox-group', className);
    var children = this.props.children;

    if (data) {
      children = data.map(function (_ref, index) {
        var text = _ref.text,
            value = _ref.value,
            rest = _objectWithoutProperties(_ref, ['text', 'value']);

        return React.createElement(
          Checkbox,
          _extends({ key: index, value: value }, rest),
          text
        );
      });
    }

    return React.createElement(
      'div',
      { className: classes },
      children
    );
  };

  return CheckboxGroup;
}(PureComponent);

CheckboxGroup.displayName = 'AUICheckboxGroup';
CheckboxGroup.propTypes = {
  disabled: bool,
  size: string,
  name: string,
  className: string,
  onChange: func,
  data: arrayOf(shape({
    text: string.isRequired,
    value: any.isRequired // eslint-disable-line
  })),
  children: node,
  value: array, // eslint-disable-line
  defaultValue: array // eslint-disable-line
};
CheckboxGroup.defaultProps = {
  onChange: identity
};
CheckboxGroup.childContextTypes = {
  auiCheckboxGroup: object
};

var _initialiseProps = function _initialiseProps() {
  var _this2 = this;

  this.onChange = function (e, _ref2) {
    var optionValue = _ref2.value;

    // Should construct one new array for re-rendering
    var value = [].concat(_this2.state.value);
    var onChange = _this2.props.onChange;

    var index = value.indexOf(optionValue);

    if (index === -1) {
      // If NOT found, then add it
      value.push(optionValue);
    } else {
      // If found, then delete it
      value.splice(index, 1);
    }

    if (!('value' in _this2.props)) {
      _this2.setState({
        value: value
      });
    }

    onChange(value);
  };
};

export default CheckboxGroup;