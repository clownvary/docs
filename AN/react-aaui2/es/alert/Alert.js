import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { PureComponent } from 'react';
import { bool, string, func, object, node, oneOf } from 'prop-types';
import classNames from 'classnames';

var icons = {
  success: 'check-circle',
  warning: 'exclamation-circle',
  danger: 'times-circle',
  error: 'times-circle',
  info: 'info-circle'
};

var Alert = function (_PureComponent) {
  _inherits(Alert, _PureComponent);

  function Alert() {
    _classCallCheck(this, Alert);

    return _possibleConstructorReturn(this, _PureComponent.apply(this, arguments));
  }

  Alert.prototype.render = function render() {
    var _classNames;

    var _props = this.props,
        className = _props.className,
        style = _props.style,
        type = _props.type,
        noClose = _props.noClose,
        inverse = _props.inverse,
        onClose = _props.onClose,
        children = _props.children,
        rest = _objectWithoutProperties(_props, ['className', 'style', 'type', 'noClose', 'inverse', 'onClose', 'children']);

    var classes = classNames((_classNames = {
      alert: true
    }, _classNames['alert-' + type] = type, _classNames['alert-' + type + '--inverse'] = type && inverse, _classNames['alert-dismissable'] = !noClose, _classNames), className);

    return React.createElement(
      'div',
      _extends({}, rest, { className: classes, style: style, role: 'alert' }),
      React.createElement('span', { className: 'icon-' + icons[type] }),
      children,
      noClose ? undefined : React.createElement(
        'button',
        { type: 'button', className: 'close', onClick: onClose },
        '\xD7'
      )
    );
  };

  return Alert;
}(PureComponent);

Alert.displayName = 'AUIAlert';
Alert.propTypes = {
  type: oneOf(['success', 'warning', 'danger', 'error', 'info']).isRequired, // `error` is an alias for `danger`
  className: string,
  style: object, // eslint-disable-line
  noClose: bool,
  inverse: bool,
  onClose: func,
  children: node
};
Alert.defaultProps = {
  type: 'info',
  noClose: false,
  inverse: false
};
export default Alert;