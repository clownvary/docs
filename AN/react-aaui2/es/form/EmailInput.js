import _extends from 'babel-runtime/helpers/extends';
import React from 'react';
import { string } from 'prop-types';

import TextInput from './TextInput';

var propTypes = { rules: string };
var EmailInput = function EmailInput(props) {
  return React.createElement(TextInput, _extends({ type: 'email' }, props, { rules: props.rules + '|email' }));
};

EmailInput.propTypes = propTypes;
EmailInput.displayName = 'EmailInput';

export default EmailInput;