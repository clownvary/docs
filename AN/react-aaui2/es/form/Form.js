import _extends from 'babel-runtime/helpers/extends';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import _Object$keys from 'babel-runtime/core-js/object/keys';

var _Form$contextTypes;

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import merge from 'lodash/merge';

import HField from './HField';
import TextInput from './TextInput';
import Store from './store';
import { noop, identity, shallowEqual, omit } from '../shared/utils';
import { FormPropTypes, FormStorePropTypes } from './types';
import { reduceFields, reduceFieldsProp } from './utils';

var aauiFormStore = _Object$keys(FormStorePropTypes)[0];
var FormDefaultConfig = {
  preSubmit: noop,
  shouldSubmit: function shouldSubmit() {
    return true;
  },
  onSubmit: noop,
  postSubmit: noop,
  onFail: noop,
  preChange: identity,
  onChange: noop,
  fields: {}
};

var Form = function (_PureComponent) {
  _inherits(Form, _PureComponent);

  function Form(props, context) {
    _classCallCheck(this, Form);

    // Sync the `this.state` and `this.store.currentState`
    var _this = _possibleConstructorReturn(this, _PureComponent.call(this, props, context));

    _this.handleSubmit = function (e) {
      e && e.preventDefault && e.preventDefault();

      _this.store.tryInvokeValidators();

      var _this$props = _this.props,
          preSubmit = _this$props.preSubmit,
          shouldSubmit = _this$props.shouldSubmit,
          onSubmit = _this$props.onSubmit,
          postSubmit = _this$props.postSubmit,
          onFail = _this$props.onFail;

      var _reduceFields = reduceFields(_this.store.getState()),
          values = _reduceFields.values,
          errors = _reduceFields.errors,
          isValid = _reduceFields.isValid;

      if (!isValid) {
        return onFail(errors);
      }

      if (shouldSubmit(values, preSubmit(values))) {
        onSubmit(values);
      }

      return postSubmit(values);
    };

    _this.handleChange = function () {
      var fields = _this.store.getState();

      var _reduceFields2 = reduceFields(fields),
          values = _reduceFields2.values,
          errors = _reduceFields2.errors,
          isValid = _reduceFields2.isValid;

      var _reduceFields3 = reduceFields(_this.state.fields),
          currentValues = _reduceFields3.values,
          currentErrors = _reduceFields3.errors;

      var _this$props2 = _this.props,
          preChange = _this$props2.preChange,
          onChange = _this$props2.onChange;


      if (!shallowEqual(currentValues, values) || !shallowEqual(currentErrors, errors)) {
        // trigger 'onChange' here on purpose
        onChange({ values: preChange(values), errors: errors, fields: fields });

        // store -> state
        _this.setState({ fields: fields, isValid: isValid, changed: true });
      }
    };

    _this.parentStore = props[aauiFormStore] || context[aauiFormStore];
    _this.state = _this.getInitState();
    _this.initStore(props);
    return _this;
  }

  Form.prototype.getChildContext = function getChildContext() {
    var _ref;

    return _ref = {}, _ref[aauiFormStore] = this.store, _ref;
  };

  Form.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    if (nextProps.fields !== this.props.fields) {
      // Merge fields when receiving new fields
      var fields = merge({}, this.store.getState(), reduceFieldsProp(nextProps.fields));
      this.store.replaceState(fields);
      this.setState({ fields: fields, changed: false });
    }
  };

  Form.prototype.componentWillUnmount = function componentWillUnmount() {
    if (this.unsubscribe) {
      this.unsubscribe();
      this.unsubscribe = null;
    }
  };

  Form.prototype.getInitState = function getInitState() {
    var fields = this.props.fields;


    return { fields: reduceFieldsProp(fields), isValid: true, changed: false };
  };

  Form.prototype.initStore = function initStore(props) {
    this.store = new Store(_extends({}, props, {
      initialState: this.state.fields
    }));
    // Subscribe the changes of store
    this.unsubscribe = this.store.subscribe(this.handleChange);
    // Keep the child store reference from the parent store
    if (this.parentStore) {
      this.parentStore.childStoreList = [].concat(this.parentStore.childStoreList, [this.store]);
    }
  };

  Form.prototype.render = function render() {
    var _this2 = this;

    var _props$className = this.props.className,
        className = _props$className === undefined ? 'form--horizontal' : _props$className;

    var children = this.props.children;

    // Data-Driven Form
    var fields = _Object$keys(this.props.fields).map(function (fieldKey, index) {
      var field = _this2.props.fields[fieldKey] || {};
      var _field$layout = field.layout,
          Layout = _field$layout === undefined ? HField : _field$layout;


      return React.createElement(Layout, _extends({
        key: index,
        component: TextInput,
        name: fieldKey
      }, omit(field, ['layout'])));
    });

    if (typeof children === 'function') {
      children = children(_extends({}, this.state, { fields: fields }));
    }

    // If we have the `this.parentStore` then it means we're the nested form
    if (this.parentStore) {
      return React.createElement(
        'div',
        null,
        children
      );
    }

    return React.createElement(
      'form',
      { className: className, onSubmit: this.handleSubmit },
      children
    );
  };

  return Form;
}(PureComponent);

/**
 * Provide utility function and lifecycle methods for `Form`
 */
/* eslint-disable react/no-multi-comp */


Form.displayName = 'AUIForm';
Form.propTypes = _extends({}, FormPropTypes);
Form.defaultProps = _extends({}, FormDefaultConfig);
Form.contextTypes = (_Form$contextTypes = {}, _Form$contextTypes[aauiFormStore] = PropTypes.instanceOf(Store), _Form$contextTypes);
Form.childContextTypes = _extends({}, FormStorePropTypes);
export default Form;
export var createForm = function createForm() {
  var formConfig = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return function (WrappedComponent) {
    var Connect = function (_PureComponent2) {
      _inherits(Connect, _PureComponent2);

      function Connect() {
        _classCallCheck(this, Connect);

        return _possibleConstructorReturn(this, _PureComponent2.apply(this, arguments));
      }

      Connect.prototype.render = function render() {
        return React.createElement(WrappedComponent, this.props);
      };

      return Connect;
    }(PureComponent);

    Connect.defaultProps = _extends({}, FormDefaultConfig, formConfig);

    return Connect;
  };
};