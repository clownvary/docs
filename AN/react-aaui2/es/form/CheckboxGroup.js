import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { PureComponent } from 'react';
import { node } from 'prop-types';

import connectForm from './connectForm';
import { noop, omit } from '../shared/utils';
import { CheckboxGroup } from '../Checkbox';
import { FormFieldAPIPropTypes } from './types';

var FormCheckboxGroup = function (_PureComponent) {
  _inherits(FormCheckboxGroup, _PureComponent);

  function FormCheckboxGroup() {
    var _temp, _this, _ret;

    _classCallCheck(this, FormCheckboxGroup);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, _PureComponent.call.apply(_PureComponent, [this].concat(args))), _this), _this.handleChange = function (value) {
      var _this$props = _this.props,
          setValue = _this$props.api.setValue,
          onChange = _this$props.onChange;


      setValue(value);
      onChange(value);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  FormCheckboxGroup.prototype.render = function render() {
    var _props = this.props,
        children = _props.children,
        rest = _objectWithoutProperties(_props, ['children']);

    return React.createElement(
      CheckboxGroup,
      _extends({}, omit(rest, ['api', 'rules']), {
        onChange: this.handleChange
      }),
      children
    );
  };

  return FormCheckboxGroup;
}(PureComponent);

FormCheckboxGroup.displayName = 'AUICheckboxGroup';
FormCheckboxGroup.propTypes = _extends({}, FormFieldAPIPropTypes, {
  children: node
});
FormCheckboxGroup.defaultProps = {
  onChange: noop
};


export default connectForm()(FormCheckboxGroup);