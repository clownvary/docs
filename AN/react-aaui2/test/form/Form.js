import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'

import L10nProvider from 'shared/L10nProvider'
import Form from 'form'
import CheckboxGroup from 'form/CheckboxGroup'
import TextInput from 'form/TextInput'
import EmailInput from 'form/EmailInput'
import Button from 'Button'
import Alert from 'alert'
import formMountOptions from '../__mocks__/formMountOptions'

const name = 'field'
const setup = (props = {}) => {
  const FieldComponent = props.FieldComponent || TextInput
  const Component = (
    <Form {...props}>
      <Form.HField
        name={name}
        label="field"
        component={FieldComponent}
        required={props.required}
        value={props.value || ''}
      />
      <Button>Submit</Button>
    </Form>
  )

  let wrapper = mount(Component)

  if (props.parentStore) {
    wrapper = mount(Component, formMountOptions)
  }

  return {
    props,
    Component,
    wrapper,
  }
}

describe('Form', () => {
  it('should render withoug errors', () => {
    const { wrapper } = setup()

    expect(wrapper).toBeTruthy()

    wrapper.unmount()
  })

  it('should render correct thing', () => {
    const { wrapper } = setup({ onSubmit: () => {} })

    expect(typeof wrapper.find(Form).props().onSubmit === 'function').toBe(true)
    expect(wrapper.find(Form.HField).length).toBe(1)
    expect(wrapper.find('button').length).toBe(1)
  })

  it("should render right thing and don't change unexpected", () => {
    const { Component } = setup({ onSubmit: () => {}, required: true })
    const tree = renderer.create(Component).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("should render right thing and don't change unexpected when nesting Form", () => {
    const { Component } = setup({
      onSubmit: () => {},
      required: true,
      parentStore: true,
    })
    const tree = renderer.create(Component).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('set default values for form fields through `defalutValue`', () => {
    const defaultValue = 'defaultValue'
    const fields = { [name]: { defaultValue } }
    const { wrapper } = setup({ fields })

    expect(
      wrapper
        .find(`[name="${name}"]`)
        .last()
        .prop('value'),
    ).toBe(defaultValue)
  })

  it('set default values for form fields directly', () => {
    const defaultValue = 'defaultValue'
    const fields = { [name]: defaultValue }
    const { wrapper } = setup({ fields })

    expect(
      wrapper
        .find(`[name="${name}"]`)
        .last()
        .prop('value'),
    ).toBe(defaultValue)
  })

  it('createForm', () => {
    const MyForm = Form.createForm()(props => (
      <L10nProvider>
        <Form {...props}>
          <Form.HField
            name="field"
            label="field"
            component={TextInput}
            required={props.required}
          />
          <button type="submit">Submit</button>
        </Form>
      </L10nProvider>
    ))

    const wrapper = mount(<MyForm required />)

    expect(wrapper).toBeTruthy()
  })

  it('EmailInput validation', () => {
    const { wrapper } = setup({ FieldComponent: EmailInput })
    const input = wrapper.find('input')
    const test = 'test'

    input.simulate('change', { target: { value: test } })
    // input.getElement().value = test
    input.simulate('blur', input)

    expect(wrapper.find('.form__group--error').length).toBeGreaterThanOrEqual(1)
    expect(
      wrapper
        .find('.form__validate')
        .last()
        .text(),
    ).toBe('This field is invalid.')
  })
})

describe('Form API', () => {
  it('lifecycle methods', () => {
    const preSubmitMock = jest.fn()
    const shouldSubmitMock = jest.fn()
    const onSubmitMock = jest.fn()
    const postSubmitMock = jest.fn()
    const onFailMock = jest.fn()
    const onChangeMock = jest.fn()
    const { wrapper } = setup({
      required: false,
      preSubmit: preSubmitMock,
      shouldSubmit: shouldSubmitMock,
      onSubmit: onSubmitMock,
      postSubmit: postSubmitMock,
      onFail: onFailMock,
      onChange: onChangeMock,
    })

    preSubmitMock.mockReturnValue('preSubmit')
    shouldSubmitMock.mockReturnValue(true)

    wrapper.simulate('submit')

    expect(preSubmitMock).toBeCalled()
    expect(shouldSubmitMock).toBeCalled()
    expect(shouldSubmitMock).toBeCalledWith({ [name]: undefined }, 'preSubmit')
    expect(onSubmitMock).toBeCalled()
    expect(postSubmitMock).toBeCalled()
    expect(onChangeMock).not.toBeCalled()

    shouldSubmitMock.mockReturnValue(false)

    wrapper.simulate('submit')
  })

  it('submit', () => {
    const onSubmitMock = jest.fn()
    const onFailMock = jest.fn()
    const onChangeMock = jest.fn()
    const { wrapper } = setup({
      required: true,
      onSubmit: onSubmitMock,
      onFail: onFailMock,
      onChange: onChangeMock,
    })
    // "Event propagation is not supported". I assume that's what causes this problem.
    // See https://github.com/airbnb/enzyme/blob/master/docs/future.md
    // wrapper.find('button[type="submit"]').simulate('click')
    wrapper.simulate('submit')

    expect(onSubmitMock.mock.calls.length).toBe(0)
    expect(onFailMock.mock.calls.length).toBe(1)
    expect(onChangeMock.mock.calls.length).toBe(1)

    const input = wrapper.find('input')
    const test = 'test'
    // Trigger `setValue`
    input.simulate('change', { target: { value: test } })
    // input.getElement().value = test
    // Trigger `onValidate`
    input.simulate('blur', input)

    expect(wrapper.find(Alert).length === 0).toBe(true)

    wrapper.simulate('submit')
    wrapper.simulate('submit')
    wrapper.simulate('submit')
    wrapper.simulate('submit')

    expect(onChangeMock.mock.calls.length).toBeGreaterThanOrEqual(3)
    expect(onSubmitMock.mock.calls.length).toBe(4)
    expect(onSubmitMock.mock.calls[0][0]).toMatchObject({ [name]: test })
  })

  it('call `parser` func before triggering validation', () => {
    const parserMock = jest.fn()
    const fields = { [name]: { parser: parserMock } }
    const { wrapper } = setup({ fields, required: true })

    const input = wrapper.find('input')
    // Trigger validation
    input.simulate('blur', input)

    expect(parserMock).toBeCalled()

    const test = 'test'
    // Trigger `setValue`
    input.simulate('change', {
      target: {
        value: test,
      },
    })
    // input.getElement().value = test

    // Trigger validation
    input.simulate('blur', input)
    expect(parserMock).toHaveBeenCalledTimes(2)
    expect(parserMock).toBeCalledWith(test, expect.anything())
  })

  it('call `formatter` func if the value is valid', () => {
    const formatterMock = jest.fn()
    const fields = { [name]: { formatter: formatterMock } }
    const { wrapper } = setup({ fields, required: true })

    const input = wrapper.find('input')
    // Trigger validation
    input.simulate('blur', input)

    expect(formatterMock).toBeCalled()

    const test = 'test'
    // Trigger `setValue`
    input.simulate('change', { target: { value: test } })
    // input.getElement().value = test

    // Trigger validation
    input.simulate('blur', input)

    expect(formatterMock).toHaveBeenCalledTimes(2)
    expect(formatterMock).toBeCalledWith(test, expect.anything())
  })
})

describe('Form fields', () => {
  it('should accept field value with type of array', () => {
    const onSubmit = jest.fn()
    const value = ['tennis', 'basketball']
    const fields = { [name]: value }
    const { wrapper } = setup({
      fields,
      onSubmit,
      FieldComponent: CheckboxGroup,
    })

    wrapper.simulate('submit')

    expect(onSubmit).toBeCalledWith(expect.objectContaining({ [name]: value }))
  })

  it('update', () => {
    const defaultValue = 'defaultValue'
    const updatedValue = `${defaultValue}--updated`
    const fields = { [name]: { defaultValue } }
    const { wrapper } = setup({ fields, needL10nProvider: true })

    expect(
      wrapper
        .find(`[name="${name}"]`)
        .last()
        .prop('value'),
    ).toBe(defaultValue)

    const updatedProps = { fields: { [name]: updatedValue } }
    wrapper.setProps(updatedProps)
    expect(
      wrapper
        .find(`[name="${name}"]`)
        .last()
        .prop('value'),
    ).toBe(updatedValue)

    wrapper.setProps(updatedProps)
    expect(
      wrapper
        .find(`[name="${name}"]`)
        .last()
        .prop('value'),
    ).toBe(updatedValue)
  })

  it('clear', () => {
    const defaultValue = 'defaultValue'
    const updatedValue = ''
    const fields = { [name]: { defaultValue } }
    const { wrapper } = setup({ fields, needL10nProvider: true })

    expect(
      wrapper
        .find(`[name="${name}"]`)
        .last()
        .prop('value'),
    ).toBe(defaultValue)

    const updatedProps = { fields: { [name]: updatedValue } }
    wrapper.setProps(updatedProps)
    expect(
      wrapper
        .find(`[name="${name}"]`)
        .last()
        .prop('value'),
    ).toBe(updatedValue)
  })
})
