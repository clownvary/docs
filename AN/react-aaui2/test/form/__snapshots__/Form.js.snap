// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render right thing and don't change unexpected 1`] = `
<form
  className="form--horizontal"
  onSubmit={[Function]}
>
  <div
    className="row row-gutter row-align-center row-justify-start form__group form__group--horizontal"
    style={Object {}}
  >
    <div
      className="col col-3 form__label form__label--require"
    >
      <label>
        <span>
          field
        </span>
      </label>
    </div>
    <div
      className="col col-6 form__control"
    >
      <div
        className="input-group input-group--m"
        style={undefined}
      >
        <input
          className="input input-group__field"
          disabled={undefined}
          label="field"
          name="field"
          onBlur={[Function]}
          onChange={[Function]}
          onInput={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <button
    className="btn btn-secondary"
    type="submit"
  >
    <span>
      Submit
    </span>
  </button>
</form>
`;

exports[`Form should render right thing and don't change unexpected when nesting Form 1`] = `
<form
  className="form--horizontal"
  onSubmit={[Function]}
>
  <div
    className="row row-gutter row-align-center row-justify-start form__group form__group--horizontal"
    style={Object {}}
  >
    <div
      className="col col-3 form__label form__label--require"
    >
      <label>
        <span>
          field
        </span>
      </label>
    </div>
    <div
      className="col col-6 form__control"
    >
      <div
        className="input-group input-group--m"
        style={undefined}
      >
        <input
          className="input input-group__field"
          disabled={undefined}
          label="field"
          name="field"
          onBlur={[Function]}
          onChange={[Function]}
          onInput={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <button
    className="btn btn-secondary"
    type="submit"
  >
    <span>
      Submit
    </span>
  </button>
</form>
`;
